### DISCUSSION TITLE
Automate moving issues based on changes to pull requests and other information

### DISCUSSION THREAD

**Discussion by azenMatt** [community/community#4254] [Projects and Issues]
When a pull request is added to an issue, or when specific field value changes (like the `Status` changes from `Shaping` to `Ready for dev`), I want to automatically do something, like assign the issue to someone on my team to review. Using the graphQL API, it's not currently possible to listen for changes to specific fields that could then trigger additional updates to am item in a Project.

We're considering adding a set of webhooks that would allow you to write workflow style Actions to automate some planning and tracking tasks. 

What would you automate in your Project? Are there any other triggers that would be useful to you and your team?

**Comment by r002** [comment]
Hi @azenMatt!  Thanks for opening this thread and soliciting feedback.  I just started using GitHub's REST API two months ago and integrating with it has been a ton of fun.  For me, GitHub's CMS is unparalleled-- leagues beyond anything in WordPress, Blogger, etc.  What I especially love about GitHub is how it bills itself (and feels like) a "developer company".  Especially with its exposing of webhooks and public APIs.

To your question-- I want to just second your thought:  Assuming I've read the [documentation](https://docs.github.com/en/github-ae@latest/developers/webhooks-and-events/webhooks/webhook-events-and-payloads#issues) correctly, I feel like GitHub's current impl of webhooks doesn't currently sport sufficiently granularity on `Issues` events.  Eg.  You guys expose a **_single_** `Issues` event that we can listen to and act on:

![image](https://user-images.githubusercontent.com/45280066/124318002-04d20b00-db46-11eb-8d21-c7d7c2d11091.png)

But in the workflow I've set up, the main drivers of how I've automated things is with the usages of `labels`.  For example, for GH `Issue Templates`, you guys don't support creating an issue with a milestone.  So in my workflow, I create a label with a `daily accomplishment` label, listen for that event, and then update the issue with a `Daily Accomplishment` milestone.

Eg.  Writing [this code](https://github.com/r002/storyline-api/blob/e7f9780760fdea4769c7010548b1f2f8f695cf26/cmd/web/main.go#L199-L209) and [this code](https://github.com/r002/storyline-api/blob/e7f9780760fdea4769c7010548b1f2f8f695cf26/ghservices/dataservice.go#L60-L83) to do this:

![image](https://user-images.githubusercontent.com/45280066/124319146-d2291200-db47-11eb-9359-5025e5caa2ac.png)

These past few months, it looks like a ton of very visible UX changes (Task Issues, Project Boards/Tables, Co-Pilot) have been landing in GitHub world.  And though many of these are neat, I feel it'd also be awesome if the API and Webhooks side of GH also got more appreciable love too.  

In short-- My humble 2c:  There are a ton of workflows out there.  Just even how I personally use Git changes drastically depending if I'm working on a personal project or contributing a PR to something open-source.  Given that, I honestly think it's impossible to build GUI-friendly tools for every use case under the sun.  It'd be like Leonidas and his merry band at Thermopylae-- a valiant effort but one that ultimately goes down in flames.

Instead-- I think it'd be cool to focus more uplift on increasing the granularity of Webhook events and the GH public API.  

### Ie. _For `Issue` events: Only fire webhook events on certain labels so we don't need to listen to all issue changes._

This will save a lot of network traffic emanating from the GH-side and also spare my poor receiver on my side the need to not process bogus requests that I don't need to listen to.  Eg.  In the current impl, I _**must**_ handle `milestone-changed" and all "labeled"/"unlabled" events for **_all_** issues.  So I get delivery chains like:

![image](https://user-images.githubusercontent.com/45280066/124320258-8bd4b280-db49-11eb-8121-edfdb2270b97.png)

and also have to dismiss any irrelevant issues with payloads like:

![image](https://user-images.githubusercontent.com/45280066/124320336-b0c92580-db49-11eb-9e56-0b43a1c9028f.png)

My poor GCP instance!  😥

GitHub has always been a developer-centric company so I feel **_leaning_** into that ethos is the way.  Spinning up a GCP/AWS free-tier instance is dead-simple nowadays and any beginner can easily read [GitHub's docs](https://docs.github.com/en/github-ae@latest/developers/webhooks-and-events/webhooks/creating-webhooks) (though, pretty Ruby-centric!  No Go love! ☹) and set up a [`GitHub automation bot`](https://docs.github.com/en/get-started/learning-about-github/types-of-github-accounts) to get started.  Empower us with more **_granular APIs_**; give us more tools to fish with!  😀

**Reply by azenMatt** [reply]
Hi @r002 - thanks for the detailed feedback here, as mentioned most of the changes we just released are front-end focused but we are looking into how we can make sure the API keeps up with it. 

I think a Webhooks dedicated to label changes (with granularity) would help a lot of use-case. Nothing coming here immediately but i'l keep you posted.

**Comment by marracuene** [comment]
I think that triggers based on existing GH concepts, without needing webhooks, would have some utility:
 - on one or more specific labels being applied to issue
 - on issue having a milestone set or changed
 - on issue changing its assignee
 - on issue having a review-request for associated PR

**Comment by gr2m** [comment]
Hi @azenMatt,

I was looking for previous feedback discussions on the lack of events for the new issues, and it sounds like this is the right place to ask? Let me know if you want me to create a separate discussion instead.

A use case I'm working on is a GitHub organization wide coordinated changelog. There are many repositories for different products and services. I'd like to enforce a flow where

1. An open PR must be assigned to the "Changelog" project
2. A custom field in the "Changelog" must be set

I can query for these things and set a pull request accordingly, but right now I don't think there are events for

- A pull request was added to / removed from a project
- a project field was edited on a pull request

Are you planning on adding webhooks / GitHub Action triggers for these events?

**Comment by probitcarwyn** [comment]
Related on the public roadmap here:

- https://github.com/github/roadmap/issues/281
